File: epi_judge_python_solutions/two_sum.py
from typing import List

from test_framework import generic_test


def has_two_sum(A: List[int], t: int) -> bool:

    i, j = 0, len(A) - 1

    while i <= j:
        if A[i] + A[j] == t:
            return True
        elif A[i] + A[j] < t:
            i += 1
        else:  # A[i] + A[j] > t.
            j -= 1
    return False


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('two_sum.py', 'two_sum.tsv',
                                       has_two_sum))

File: epi_judge_cpp_solutions/two_sum.cc
#include <iterator>
#include <vector>

#include "test_framework/generic_test.h"

using std::size;
using std::vector;

bool HasTwoSum(const vector<int>& A, int t) {
  int i = 0, j = size(A) - 1;
  while (i <= j) {
    if (A[i] + A[j] == t) {
      return true;
    } else if (A[i] + A[j] < t) {
      ++i;
    } else {  // A[i] + A[j] > t.
      --j;
    }
  }
  return false;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"A", "t"};
  return GenericTestMain(args, "two_sum.cc", "two_sum.tsv", &HasTwoSum,
                         DefaultComparator{}, param_names);
}
